## Importing the Libraries

In [17]:
# Importing the Libraries
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

## Data Description







**Data Description
books — livros:**

Contains data about books:

- book_id — book identifier
- author_id — author identifier
- title — title
- num_pages — number of pages
- publication_date — publication date
- publisher_id — publisher identifier

**authors — authors:**

Contains data about authors:

- author_id —  author identifier
- author — author
- publishers — publishers:

**Contains data about publishers:**

- publisher_id —  publisher identifier
- publisher — publisher name
- ratings — ratings:

**Contains data about user ratings:**

- rating_id — rating identifier
- book_id —  book identifier
- username — the username of the person who rated the book
- rating — rating
- reviews — reviews:

**Contains data about customer reviews:**

- review_id — review identifier
- book_id — book identifier
- username — the username of the person who reviewed the book
- text — review text

## Connecting to the Database

In [18]:
# Connection Settings
db_config = {
    'user': 'practicum_student',  # username
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',  # password
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',  # host
    'port': 5432,  # port
    'db': 'data-analyst-final-project-db'  # database name
}

# Connection String
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

# The connection is stored in the variable 'engine'
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [19]:
# pd.io.sql.read_sql(query, con = engine)

## Displaying the Tables

In [20]:
query = """
SELECT *
FROM books
LIMIT 10;
"""

df_books_ex = pd.io.sql.read_sql(query, con = engine)

print(df_books_ex)

   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-1

In [21]:
query = """
SELECT *
FROM reviews
LIMIT 10;
"""
df_reviews_ex = pd.io.sql.read_sql(query, con = engine)

print(df_reviews_ex)

   review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game p

In [22]:
query = """
SELECT *
FROM publishers
LIMIT 10;
"""

df_publishers_ex = pd.io.sql.read_sql(query, con = engine)

print(df_publishers_ex)

   publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers


In [23]:
query = """
SELECT *
FROM ratings
LIMIT 10;
"""
df_ratings_ex = pd.io.sql.read_sql(query, con = engine)

print(df_ratings_ex)

   rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4


In [24]:
query = """
SELECT *
FROM ratings
LIMIT 10;
"""
df_ratings_ex = pd.io.sql.read_sql(query, con = engine)

print(df_ratings_ex)

   rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4


# Tasks

- Find the number of books released after January 1, 2000.
- Find the number of ratings and the average rating for each book.
- Identify the publisher that released the most books with more than 50 pages (this will help exclude pamphlets and similar publications from your analysis).
- Identify the author with the highest average book rating: consider only books with at least 50 ratings.
- Find the average number of ratings among users who have rated more than 50 books.

In [25]:
# Find the number of books released after January 1, 2000.
query = """
SELECT COUNT(*) AS num_books
FROM books
WHERE publication_date > '2000-01-01';
"""
df_num_books = pd.io.sql.read_sql(query, con = engine)

print(df_num_books)

   num_books
0        819


Note: A total of 819 books were found to have been released after January 1, 2000.

In [26]:
# Find the number of ratings and the average rating for each book.
query = """
SELECT book_id, COUNT(*) AS num_rating, AVG(rating) AS avg_rating
FROM ratings
GROUP BY book_id;
"""

df_ratings = pd.io.sql.read_sql(query, con = engine)

print(df_ratings)

     book_id  num_rating  avg_rating
0        652           2    4.500000
1        273           2    4.500000
2         51          12    4.250000
3        951           2    4.000000
4        839           7    4.285714
..       ...         ...         ...
995       64          13    4.230769
996       55           2    5.000000
997      148           7    3.428571
998      790           2    3.500000
999      828           2    3.000000

[1000 rows x 3 columns]


Notes:

Books with fewer reviews, such as 2 reviews, tend to have extreme ratings (such as 5.0 or 3.0). This means that the volume of reviews can influence the stability of the average rating.

Books with a moderate number of reviews, between 7 and 13 reviews, show a greater variety of ratings. This may indicate a more diverse opinion from readers.

Some books with few reviews achieve perfect ratings, such as the book with ID 55, which has a 5.0 rating based on only 2 reviews.

Conclusion after observations: We can conclude that books with a higher number of reviews provide a more reliable perspective for readers, while books with few reviews should be analyzed carefully, as they may not represent the opinion of the majority.

In [27]:
# Identify the publisher that released the most books with more than 50 pages (this will help exclude pamphlets and similar publications from your analysis).
query = """
SELECT publishers.publisher, COUNT(*) AS num_books
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY num_books DESC
LIMIT 1;
"""

df_publishers = pd.io.sql.read_sql(query, con = engine)

print(df_publishers)

       publisher  num_books
0  Penguin Books         42


Note: Penguin Books was the publisher that released the most books with over 50 pages.

In [28]:
# Identify the author with the highest average book rating: consider only books with at least 50 ratings.
query = """
WITH filtered_books AS (
    SELECT books.book_id, books.author_id, AVG(ratings.rating) AS avg_book_rating, COUNT(ratings.rating) AS num_ratings
    FROM books
    JOIN ratings ON books.book_id = ratings.book_id
    GROUP BY books.book_id, books.author_id
    HAVING COUNT(ratings.rating) >= 50
)

SELECT authors.author, AVG(filtered_books.avg_book_rating) AS avg_author_rating
FROM filtered_books
JOIN authors ON filtered_books.author_id = authors.author_id
GROUP BY authors.author
ORDER BY avg_author_rating DESC
LIMIT 1;
"""

df_top_author = pd.io.sql.read_sql(query, con=engine)

print(df_top_author)

                       author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844


Note: J.K. Rowling/Mary GrandPré has the highest average book ratings with at least 50 ratings, meaning J.K. Rowling/Mary GrandPré is a highly rated author, as their books maintain an average of 4.28.

In [29]:
# Find the average number of ratings among users who have rated more than 50 books.
query = """
WITH users_with_more_than_50_ratings AS (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(book_id) > 50
)
SELECT AVG(review_count) AS avg_review_per_user
FROM (
    SELECT reviews.username, COUNT(reviews.review_id) AS review_count
    FROM reviews
    JOIN users_with_more_than_50_ratings ON reviews.username = users_with_more_than_50_ratings.username
    GROUP BY reviews.username
) AS user_review_counts;
"""

df_users_50_ratings = pd.io.sql.read_sql(query, con=engine)

print(df_users_50_ratings)

   avg_review_per_user
0            24.333333


Note: With an average of 24.33 ratings per user, we can observe that after reaching the threshold of 24 reviews, users tend to stop rating a larger number of books.

### Conclusion:

- After the analysis, it can be concluded that books with more reviews provide a more balanced, reliable, and representative view of readers' opinions.

- Publishers that release more books, such as Penguin Books, demonstrate a strong market presence, which may be related to their popularity and influence.

- J.K. Rowling/Mary GrandPré stands out as an author with a high average rating, being well-regarded and with books that hold high credibility among readers.

- Additionally, we can observe that after reaching a certain number of reviews, users tend to decrease or stop reviewing other books, suggesting a behavior of saturation in terms of participation.